src/WCSimPrimaryGeneratorAction.cc

```diff
@@ -123,53 +123,55 @@ WCSimPrimaryGeneratorAction::WCSimPrimaryGeneratorAction(
     inputCosmicsFile.close();
 
 
-  inputCosmicsFile.open(cosmicsFileName, std::fstream::in);
-
-  if (!inputCosmicsFile.is_open()) {
-    G4cout << "Cosmics data file " << cosmicsFileName << " not found" << G4endl;
-	exit(-1);
-  } else {
-    G4cout << "Cosmics data file " << cosmicsFileName << " found" << G4endl;
-    string line;
-    vector<string> token(1);
-
-    double binCos, binPhi;
-    //double cosThetaMean, cosThetaMin, cosThetaMax;
-    //double phiMean, phiMin, phiMax;
-    double flux;
-    double Emean;
-
-    hFluxCosmics = new TH2D("hFluxCosmics","HK Flux", 180,0,360,100,0,1);
-    hFluxCosmics->GetXaxis()->SetTitle("#phi (deg)");
-    hFluxCosmics->GetYaxis()->SetTitle("cos #theta");
-    hEmeanCosmics = new TH2D("hEmeanCosmics","HK Flux", 180,0,360,100,0,1);
-    hEmeanCosmics->GetXaxis()->SetTitle("#phi (deg)");
-    hEmeanCosmics->GetYaxis()->SetTitle("cos #theta");
-
-    while ( getline(inputCosmicsFile,line) ){
-      token = tokenize(" $", line);
-
-      binCos=(atof(token[0]));
-      binPhi=(atof(token[1]));
-      /*
-      cosThetaMean=(atof(token[2]));
-      cosThetaMin=(atof(token[3]));
-      cosThetaMax=(atof(token[4]));
-      phiMean=(atof(token[5]));
-      phiMin=(atof(token[6]));
-      phiMax=(atof(token[7]));
-      */
-      flux=(atof(token[8]));
-      Emean=(atof(token[9]));
+  if (useCosmics) {
```

Comment:
Do we agree this will be not run all the time (unless one set the useCosmics variable to true)?

Reply:
Good spot @guiguem In practice this is never true, because `useCosmics` is by default `false` and only ever becomes true via the messenger (mac command `/mygen/generator cosmics`), which happens **after** this constructor

I think this is a nice idea (I'm always confused as to why it produces the extra root file), but this block needs to be moved somewhere else. For example to, `GeneratePrimaries()` (or better, a function called from there, because `GeneratePrimaries()` is already a massive unreadable function...) with a flag to say whether this block has been run (so it only gets run once)

Reply:
Thanks @guiguem and @tdealtry for the catch. 
I added this part of the code into the `GeneratePrimaries()` throught the `Create_cosmics_histogram` function.